python/lsst/dax/imgserv/getimage/coadd.py

```diff
@@ -0,0 +1,177 @@
+            wcs=wcs,
+            badMaskPlanes=config.badMaskPlanes,
+            logName=logName,
+        )
+
+    def addExposure(self, exposure, weightFactor=1.0):
+        """Add an Exposure to the coadd
+
+        @param[in] exposure: Exposure to add to coadd; this should be:
+            - background-subtracted or background-matched to the other images being coadded
+            - psf-matched to the desired PSF model (optional)
+            - warped to match the coadd
+            - photometrically scaled to the desired flux magnitude
+        @param[in] weightFactor: extra weight factor for this exposure
+
+        @return
+        - overlapBBox: region of overlap between exposure and coadd in parent coordinates (afwGeom.Box2I)
+        - weight: weight with which exposure was added to coadd; weight = weightFactor / clipped mean variance
+
+        Subclasses may override to preprocess the exposure or change the way it is added to the coadd.
+        """
+        maskedImage = exposure.getMaskedImage()
+
+        # compute the weight
+        statObj = afwMath.makeStatistics(maskedImage.getVariance(), maskedImage.getMask(),
+                                         afwMath.MEANCLIP, self._statsControl)
+        meanVar = statObj.getResult(afwMath.MEANCLIP)[0]
+        weight = weightFactor / float(meanVar)
+        if math.isnan(weight):
+            raise RuntimeError("Weight is NaN (weightFactor=%s; mean variance=%s)" % (weightFactor, meanVar))
+
+        # save filter info
+        filter = exposure.getFilter()
+        self._filterDict.setdefault(filter.getName(), filter)
+
+        self._log.info("Add exposure to coadd with weight=%0.3g", weight)
+
+        overlapBBox = coadd_utils.addToCoadd(self._coadd.getMaskedImage(),
+                                             self._weightMap, maskedImage,
+                                             self._badPixelMask, weight)
+
+        return overlapBBox, weight
+
+    def getCoadd(self):
+        """Get the coadd exposure for all exposures you have coadded so far
+
+        If all exposures in this coadd have the same-named filter then that filter is set in the coadd.
+        Otherwise the coadd will have the default unknown filter.
+
+        @warning: the Calib is not be set.
+        """
+        # make a deep copy so I can scale it
+        coaddMaskedImage = self._coadd.getMaskedImage()
+        scaledMaskedImage = coaddMaskedImage.Factory(coaddMaskedImage, True)
+
+        # set the edge pixels
+        coadd_utils.setCoaddEdgeBits(scaledMaskedImage.getMask(),
+                                     self._weightMap)
+
+        # scale non-edge pixels by weight map
+        scaledMaskedImage /= self._weightMap
+
+        scaledExposure = afwImage.makeExposure(scaledMaskedImage, self._wcs)
+        if len(self._filterDict) == 1:
+            scaledExposure.setFilter(list(self._filterDict.values())[0])
+        return scaledExposure
+
+    def getFilters(self):
```

Comment:
You might think about doing python properties for this, with a decorator @property

Reply:
This file was inherited from coadd_utils in LSST stack that John Swinbank has since deleted in approved RFC, not knowing imgserv using it.  Will need to decorate it nicely for ownership inside imgserv.